data/models.py

```diff
@@ -254,6 +254,15 @@ def save(self, *args, **kwargs):
         if self.stage_group is not None and self.stage_group.user != self.user:
             raise ValidationError('stage group user does not match job user')
         super(Job, self).save(*args, **kwargs)
+        if self.should_create_activity():
+            JobActivity.objects.create(job=self, state=self.state, step=self.step)
+
+    def should_create_activity(self):
+        job_activities = JobActivity.objects.filter(job=self).order_by('-created')
```

Comment:
FYI if you weren't aware, ForeignKey fields provide a reverse relationship, so you could instead write:

```
job_activities = self.job_activities.order_by('-created')
```

No need to change it, it's just a non-obvious Django models feature that comes in handy, especially for complicated relationships.